kube-controllers/pkg/controllers/node/ipam.go

```diff
@@ -1098,27 +1088,81 @@ func (c *IPAMController) garbageCollectKnownLeaks() error {
 			continue
 		}
 
-		logc.Info("Garbage collecting leaked IP address")
-		unallocated, err := c.client.IPAM().ReleaseIPs(context.TODO(), a.ReleaseOptions())
-		if _, ok := err.(cerrors.ErrorResourceDoesNotExist); ok || len(unallocated) == 1 {
-			logc.WithField("handle", a.handle).Debug("IP already released")
-			continue
-		} else if err != nil {
-			logc.WithError(err).WithField("handle", a.handle).Warning("Failed to release leaked IP")
-			return err
+		opts = append(opts, a.ReleaseOptions())
+		leaks[a.ReleaseOptions().Address] = a
+
+		if len(opts) >= maxBatchSize {
+			break
 		}
+	}
+
+	if len(opts) == 0 {
+		// Nothing to do.
+		return nil
+	}
+
+	// By releasing multiple IPs at once, we can reduce the number of API calls the underlying IPAM code needs to make
+	// in order to release the IPs. This is especially apparent when there are multple IP addresses from the same block
+	// that must be released, as they can all be released in a single API call to update the block.
+	log.WithField("num", len(opts)).Info("Garbage collecting leaked IP addresses")
+	_, releasedOpts, err := c.client.IPAM().ReleaseIPs(context.TODO(), opts...)
+
+	// First, go through the returned options and update allocation state. These are the IPs that were successfully
+	// released, or were unallocated to begin with. In either case, we can mark them as released.
+	for _, opt := range releasedOpts {
+		// Find the allocation that matches these release options.
+		a, ok := leaks[opt.Address]
+		if !ok {
+			log.WithField("opt", opt).Fatalf("BUG: unable to find allocation for release options: %+v", leaks)
+		}
+		logc := log.WithFields(a.fields())
 
 		// No longer a leak. Remove it here so we're not dependent on receiving
 		// the update from the syncer (which we will do eventually, this is just cleaner).
 		c.allocationState.release(a)
-
 		c.incrementReclamationMetric(a.block, a.node())
+		delete(c.confirmedLeaks, a.id())
+
+		logc.Info("Successfully garbage collected leaked IP address")
+		delete(leaks, opt.Address)
+	}
 
-		delete(c.confirmedLeaks, id)
+	// Note any leaks that we couldn't release.
+	for _, a := range leaks {
+		logc := log.WithFields(a.fields())
+		logc.Warn("Leaked IP address was not successfully garbage collected")
+	}
+
+	// Check the error.
+	if err != nil {
+		if _, ok := err.(cerrors.ErrorResourceDoesNotExist); !ok {
+			log.WithError(err).Warn("Failed to garbage collect one or more leaked IP addresses")
+			return err
+		}
 	}
 	return nil
 }
 
+func (c *IPAMController) releaseNodes(nodes []string) error {
+	var storedErr error
+	if len(nodes) > 0 {
+		log.WithField("num", len(nodes)).Info("Found a batch of nodes to release")
+	}
```

Comment:
I know with len(nodes) == 0, the following for loop is a no-op, but I think it's more idiomatic to have:
```
if len(nodes) == 0 {
  return nil
}
log.WithField("num", len(nodes)).Info("Found a batch of nodes to release")
```